persistence/pre_persistence/app.go

```diff
@@ -0,0 +1,311 @@
+}
+
+func (m *PrePersistenceContext) InsertApplication(address []byte, publicKey []byte, output []byte, paused bool, status int, maxRelays string, stakedTokens string, chains []string, pausedHeight int64, unstakingHeight int64) error {
+	if _, exists, _ := m.GetApp(address); exists {
+		return fmt.Errorf("already exists in world state")
+	}
+	cdc := Cdc()
+	db := m.Store()
+	key := append(AppPrefixKey, address...)
+	app := App{
+		Address:         address,
+		PublicKey:       publicKey,
+		Paused:          paused,
+		Status:          int32(status),
+		Chains:          chains,
+		MaxRelays:       maxRelays,
+		StakedTokens:    stakedTokens,
+		PausedHeight:    uint64(pausedHeight),
+		UnstakingHeight: unstakingHeight,
+		Output:          output,
+	}
+	bz, err := cdc.Marshal(&app)
+	if err != nil {
+		return err
+	}
+	return db.Put(key, bz)
+}
+
+func (m *PrePersistenceContext) UpdateApplication(address []byte, maxRelaysToAdd string, amountToAdd string, chainsToUpdate []string) error {
+	app, exists, _ := m.GetApp(address)
+	if !exists {
+		return fmt.Errorf("does not exist in world state")
+	}
+	cdc := Cdc()
+	db := m.Store()
+	key := append(AppPrefixKey, address...)
+	// compute new values
+	stakedTokens, err := StringToBigInt(app.StakedTokens)
+	if err != nil {
+		return err
+	}
+	stakedTokensToAddI, err := StringToBigInt(amountToAdd)
+	if err != nil {
+		return err
+	}
+	stakedTokens.Add(stakedTokens, stakedTokensToAddI)
+	maxRelays, err := StringToBigInt(app.MaxRelays)
+	if err != nil {
+		return err
+	}
+	maxRelaysToAddI, err := StringToBigInt(maxRelaysToAdd)
+	if err != nil {
+		return err
+	}
+	maxRelays.Add(maxRelays, maxRelaysToAddI)
+	// update values
+	app.MaxRelays = BigIntToString(maxRelays)
+	app.StakedTokens = BigIntToString(stakedTokens)
+	app.Chains = chainsToUpdate
+	// marshal
+	bz, err := cdc.Marshal(app)
+	if err != nil {
+		return err
+	}
+	return db.Put(key, bz)
+}
+
+func (m *PrePersistenceContext) DeleteApplication(address []byte) error {
+	if exists, _ := m.GetAppExists(address); !exists {
+		return fmt.Errorf("does not exist in world state")
+	}
+	db := m.Store()
+	key := append(AppPrefixKey, address...)
+	return db.Put(key, DeletedPrefixKey)
+}
+
+func (m *PrePersistenceContext) GetAppsReadyToUnstake(height int64, status int) (apps []*types.UnstakingActor, err error) { // TODO delete unstaking
+	db := m.Store()
+	unstakingKey := append(UnstakingAppPrefixKey, []byte(fmt.Sprintf("%d", height))...)
+	if has := db.Contains(unstakingKey); !has {
+		return nil, nil
+	}
+	val, err := db.Get(unstakingKey)
+	if err != nil {
+		return nil, err
+	}
+	if val == nil {
+		return make([]*types.UnstakingActor, 0), nil
+	}
+	unstakingApps := types.UnstakingActors{}
+	if err := proto.Unmarshal(val, &unstakingApps); err != nil {
+		return nil, err
+	}
+	for _, app := range unstakingApps.UnstakingActors {
+		apps = append(apps, app)
+	}
+	return
+}
+
+func (m *PrePersistenceContext) GetAppStatus(address []byte) (status int, err error) {
```

Comment:
I think status should be an enum, because as a reviewer it's hard for me to understand the range of values this can take.

Feel free to add a TODO for this.

Reply:
That doesn't Jive well with the intermodule primitive design strategy

Reply:
Makes sense to me.